# Exercise 4.5 preparation

## Import libraries

In [1]:
# import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# create path

path = r'C:\Users\leobs\OneDrive\Data Program\Achievement 4_Instacart\Instacart Basket Analysis'

## Load data

In [3]:
# Load product data set

df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
# check headings for accurate load

df_prods.head()

#appears accurate but need to remove columns

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [5]:
# remove unnamed

df_prods = df_prods.drop(columns = ['Unnamed: 0'])

In [6]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [7]:
# load orders data set from wrangled data

df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [8]:
# quick check headings for accurate load

df_ords.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id  order_number  \
0             0           0   2539329        1             1   
1             1           1   2398795        1             2   
2             2           2    473747        1             3   
3             3           3   2254736        1             4   
4             4           4    431534        1             5   

   order_day_of_week  order_hour_of_day  days_since_last_order  
0                  2                  8                    NaN  
1                  3                  7                   15.0  
2                  3                 12                   21.0  
3                  4                  7                   29.0  
4                  4                 15                   28.0

In [9]:
# remove columns

df_ords = df_ords.drop(columns = ['Unnamed: 0.1', 'Unnamed: 0'])

## Data consistency checks

In [10]:
# load df.describe for orders df

df_ords.describe()

order_id       user_id  order_number  order_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06       3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01       2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01       2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00       0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00       1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01       3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01       5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02       6.000000e+00   

       order_hour_of_day  days_since_last_order  
count       3.421083e+06           3.214874e+06  
mean        1.345202e+01           1.111484e+01  
std         4.226088e+00           9.206737e+00  
min         0.000000e+00           0.000000e+00  
25%         1.000000e+01           4.000000e+00  
50%         1.300000e+01           7.000000e+00  
75%         1.600000e+01           1.500000e+01  
max         2.300000e+01           3.000000e+01

## Mixed-type data

In [11]:
# Create a datafram

df_test = pd.DataFrame()

In [12]:
# Create a mixed data type column

df_test['mix'] = ['a', 'b', 1, True]

In [13]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [14]:
#Check for mixed types

for col in df_test.columns.tolist(): 
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1) 
    if len (df_test[weird]) > 0: 
        print (col)

mix


In [15]:
# Change data type from mix to string

df_test['mix'] = df_test['mix'].astype('str')

In [16]:
# Check for succesful data type change

df_test.dtypes

mix    object
dtype: object

## Missing values

In [17]:
# Assign function to df_prods dataframe to find number of missing observations

df_prods.isnull().sum()

#Only product_name column has missing values, and 16 observations are missing

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [18]:
# Create new dataframe only containing the 16 observations with missing values

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [19]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [20]:
# Create new dataframe only containing rows without missing data from df_prods

df_prods_clean = df_prods[df_prods['product_name'].isnull () == False]

In [21]:
# Check for removal of missing rows

df_prods_clean.shape

# Confirmed as properly removed, exactly 16 rows less

(49677, 5)

## Check for full duplicates

In [22]:
# identify all full duplicates in cleaned products df

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [23]:
# list of full duplicates 

df_dups

# there are 5 full duplicates in the df_prods_clean data frame

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

## Remove full duplicates as a new dataframe

In [24]:
# create new df that drops all duplicates

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [25]:
# check shack in df_prods_clean_no_dups to make sure only duplicates dropped, total rows minus duplicated rows

df_prods_clean_no_dups.shape

#accurate number dropped, 5 fewer

(49672, 5)

In [26]:
# Export cleaned and no duplicates products dataframe as new file

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'df_prods_checked.csv'))

# Exercise 4.5 task: consistency checks

## 1 Finalize data checks for df_prods_checked

Note: already checked for missing values (16) and full duplicates (5) and removed them from df_prods from exericse work.

Now will import updated data file 'df_prods_checked' to complete remaining data consistency checks

In [27]:
# Load products data set

df_prods_checked = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'df_prods_checked.csv'), index_col = False)

In [28]:
df_prods_checked.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [29]:
# remove unnamed columns

df_prods_checked = df_prods_checked.drop(columns = ['Unnamed: 0'])

In [30]:
df_prods_checked.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [31]:
# descriptive stats for df_prods_checked

df_prods_checked.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

All columns have the same count, showing consistency 
The product_id max does not match the count of product_ids, but the difference likely due to removal of missing values and duplicates. 
All columns have a minimum value of 1, consistent 
Min, max, and quartiles of product_id, aisle_id, and department_id appear consistent

Further investigation Max value of 99999.00 for prices seems too high, especially given quartile values

## Address inconsistencies with 'price' outliers

In [33]:
# Investigate price max of 99999.00 given high price for a grocery product, starting with price point 50

df_prods_checked.loc[df_prods_checked['prices'] > 50.00]

# returns results of two key outliers, product ids 21553 and 33664

product_id                      product_name  aisle_id  department_id  \
21538       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33649       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21538  14900.0  
33649  99999.0

In [34]:
# Check for all milk products in diary to replace milk price for product_id 3364
df_milk_prods = df_prods_checked.loc[(df_prods_checked['product_name'].str.contains ('milk')) & (df_prods_checked['department_id'] ==16)]          

In [35]:
# new df of all products containing 'milk' and in department 16, diary
df_milk_prods

product_id                                       product_name  \
133           137                               Very Vanilla Soymilk   
204           208  Vanilla Almond Cashew Almondmilk Cashewmilk Blend   
324           329                       Organic Whole Grassmilk Milk   
1396         1402                              Plain Soymilk Creamer   
2464         2473                                       West Soymilk   
...           ...                                                ...   
45986       46003                        Low Fat Cultured Buttermilk   
46658       46675                        Vanilla Cultured Almondmilk   
47746       47763                        Pure Unsweetened Cashewmilk   
48628       48645                            Unsweetened Coconutmilk   
49321       49338                          Original Enriched Soymilk   

       aisle_id  department_id  prices  
133          91             16     1.2  
204          91             16     3.2  
324          84             16     7.5  
1396         91             16    10.5  
2464         91             16     7.9  
...         ...            ...     ...  
45986        84             16    10.6  
46658       120             16    11.2  
47746        91             16     9.4  
48628        91             16     4.7  
49321        91             16     2.7  

[84 rows x 5 columns]

In [36]:
df_milk_prods['prices'].median()

#will use median value to replace milk price

7.2

In [37]:
df_milk_prods['prices'].mean()

7.517857142857144

In [38]:
# Create new df for cottage cheese products to find median price for replacement

df_cottage_c_prods = df_prods_checked.loc[df_prods_checked['product_name'].str.contains ('Cottage')]

In [39]:
df_cottage_c_prods

product_id                                       product_name  \
191           195  Grade A Pasteurized 2% Milkfat Lowfat Cottage ...   
1497         1503                             Low Fat Cottage Cheese   
2568         2577       2% Milk Fat Creamy Small Curd Cottage Cheese   
2933         2943  Low Fat, Small Curd 1% Milkfat No Salt Added C...   
3775         3788             Strawberry 100 Calorie Cottage Doubles   
...           ...                                                ...   
46770       46787                            Fat Free Cottage Cheese   
47088       47105                   Organic Pineapple Cottage Cheese   
47090       47107  4% Milkfat Small Curd Grade A Pasteurized Cott...   
47421       47438                          Pineapple Cottage Doubles   
48242       48259         Non Fat Grade A Pasteurized Cottage Cheese   

       aisle_id  department_id  prices  
191         108             16     3.3  
1497        108             16     6.7  
2568        100             21     6.5  
2933        108             16     2.2  
3775         21             16     2.1  
...         ...            ...     ...  
46770       108             16     3.4  
47088       100             21    12.6  
47090       108             16     6.4  
47421       108             16    10.8  
48242       108             16     3.7  

[78 rows x 5 columns]

In [40]:
# median price of cottage cheese products to update product_id '21553' price value

df_cottage_c_prods['prices'].median()

6.85

In [41]:
# update inaccurate outliers, replace product_id '33664' price from 99999.0000 to 7.2 and replace product_id '21553' from price of 14900.0 to 6.85

df_prods_checked = df_prods_checked.replace ({'prices':{99999.0: 7.2, 14900.0: 6.85}})

In [42]:
# check if changes executed properly for Cottage Cheese

df_prods_checked[df_prods_checked['product_id']==21553]

#yes

product_id                      product_name  aisle_id  department_id  \
21538       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   

       prices  
21538    6.85

In [43]:
# check if changes executed properly for 2 % Milk

df_prods_checked[df_prods_checked['product_id']==33664]

#yes

product_id           product_name  aisle_id  department_id  prices
33649       33664  2 % Reduced Fat  Milk        84             16     7.2

In [44]:
# rerun descriptive stats to check for consistencies with updated outlier values

df_prods_checked.describe()

# consistency check shows accurate results

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      7.680410
std    14340.705287     38.315784       5.850779      4.199299
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000     25.000000

## Check for mixed-data types in df_prods_checked

In [45]:
# check for mixed-data types

for col in df_prods_checked.columns.tolist():
    weird = (df_prods_checked[[col]].applymap(type) != df_prods_checked[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_prods_checked[weird]) > 0:
        print (col)
        
# no columns printed means no mixed-data types in this data frame

As noted, already completed checks for missing values for df_prods_checked and duplicates from above exercise

In [46]:
# Export cleaned and checked data as data file, due to replacing outliers and with completed consistency checks

df_prods_checked.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'df_prods_checked.csv'))

# Consistency checks for df_ords dataframe

## 2 Examine descriptive stats for df_ords

In [47]:
#descriptive states for df_ords

df_ords.describe()

order_id       user_id  order_number  order_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06       3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01       2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01       2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00       0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00       1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01       3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01       5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02       6.000000e+00   

       order_hour_of_day  days_since_last_order  
count       3.421083e+06           3.214874e+06  
mean        1.345202e+01           1.111484e+01  
std         4.226088e+00           9.206737e+00  
min         0.000000e+00           0.000000e+00  
25%         1.000000e+01           4.000000e+00  
50%         1.300000e+01           7.000000e+00  
75%         1.600000e+01           1.500000e+01  
max         2.300000e+01           3.000000e+01

All columns have matching number of rows except for days_since_last_order (see comment below)

order_id, user_id, and order_number columns have 1 as the min value. 
order_day_of_week, order_hour_of_day, and days_since_last_order all have a 0 as min 
order_id has a large range, matching the idea of a unique identifier for each record 
max value for user_id does not match count of user_id, referencing customers having multiple orders 

the max value for order number is less than the counts of order_numbers, referencing orders that need to investigate 
max value of days_since_last_order of 30 does not seem accurate, want to investigate further
The count of days_since_last_order does not match the counts of the other 5 columns, which means that there are significant missing values in the days_since_last_order. This could be from error or because it is a first time customer, will investigate further.

## Explore df_ords to help with descriptive stats investigation

In [48]:
# View df_ords first 5 rows
df_ords.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  
0                    NaN  
1                   15.0  
2                   21.0  
3                   29.0  
4                   28.0

In [49]:
# View df_ords last 5 rows
df_ords.tail()

order_id  user_id  order_number  order_day_of_week  \
3421078   2266710   206209            10                  5   
3421079   1854736   206209            11                  4   
3421080    626363   206209            12                  1   
3421081   2977660   206209            13                  1   
3421082    272231   206209            14                  6   

         order_hour_of_day  days_since_last_order  
3421078                 18                   29.0  
3421079                 10                   30.0  
3421080                 12                   18.0  
3421081                 12                    7.0  
3421082                 14                   30.0

In [50]:
# Check df_ords size
df_ords.shape

(3421083, 6)

## 3 Check for mixed-data types for df_ords

In [52]:
# check for mixed-data types

for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print (col)

## 4 Fix mixed-data types for df_ords

No columns printed means no mixed-data types detected for df_ords

## 5 Check for missing values

As noted in descriptive stats observations, days_since_last_order has a different count than all other columns

In [53]:
# check for missing values in df_ords

df_ords.isnull().sum()

#consistent with exploratory investigation

order_id                      0
user_id                       0
order_number                  0
order_day_of_week             0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

In [56]:
# create subset of missing values within df_ords for days_since_last_order

df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull() == True]

In [57]:
# View missing values 

df_ords_nan.head(50)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0     2539329        1             1                  2                  8   
11    2168274        2             1                  2                 11   
26    1374495        3             1                  1                 14   
39    3343014        4             1                  6                 11   
45    2717275        5             1                  3                 12   
50    2086598        6             1                  5                 18   
54    2565571        7             1                  3                  9   
75     600894        8             1                  6                  0   
79     280530        9             1                  1                 17   
83    1224907       10             1                  2                 14   
89    2411567       11             1                  0                 11   
97    1483446       12             1                  1                 12   
103   2618231       13             1                  6                 12   
116   2906263       14             1                  2                 18   
130   2514324       15             1                  5                 13   
153   1273676       16             1                  2                  8   
160   1737705       17             1                  2                 13   
201   2780889       18             1                  6                 18   
208   1321974       19             1                  6                 13   
218    947722       20             1                  1                 10   
223   2625565       21             1                  1                 10   
257   2300537       22             1                  3                  7   
273   1792521       23             1                  5                 14   
278    611214       24             1                  3                 20   
297   2610576       25             1                  1                 14   
301   3229714       26             1                  2                 15   
314    361493       27             1                  3                  9   
396   1138968       28             1                  0                 14   
421   2064090       29             1                  5                 11   
440    375303       30             1                  1                 13   
449   2150649       31             1                  5                 11   
470    134879       32             1                  3                 12   
476   2906103       33             1                  3                 14   
480   2383661       34             1                  4                 15   
486   1269089       35             1                  5                 21   
496   3285797       36             1                  4                 17   
534    327827       37             1                  2                 10   
558   2873174       38             1                  0                 10   
571   2782479       39             1                  0                 12   
579   2570028       40             1                  0                  9   
589    252244       41             1                  6                 16   
595    940313       42             1                  2                 10   
612    606677       43             1                  1                 18   
624   1359969       44             1                  0                  7   
628    300892       45             1                  1                 20   
633   3345812       46             1                  1                 16   
653   2857489       47             1                  5                 16   
659   1035233       48             1                  6                 16   
670   1216869       49             1                  2                  9   
679   1363380       50             1                  3                  9   

     days_since_last_order

In [58]:
df_ords_nan.tail(50)

order_id  user_id  order_number  order_day_of_week  \
3420283   1684894   206160             1                  2   
3420294   1013729   206161             1                  0   
3420310   2614637   206162             1                  3   
3420329    464023   206163             1                  6   
3420335   1172115   206164             1                  3   
3420341   2083902   206165             1                  3   
3420369   3290085   206166             1                  4   
3420389    351134   206167             1                  4   
3420394   2450849   206168             1                  5   
3420407   2668285   206169             1                  1   
3420416   3202089   206170             1                  1   
3420423    896963   206171             1                  2   
3420430   2526805   206172             1                  6   
3420437   1717723   206173             1                  1   
3420441    559098   206174             1                  0   
3420494   1946643   206175             1                  6   
3420501   2939252   206176             1                  6   
3420516   2441067   206177             1                  6   
3420524   2962368   206178             1                  3   
3420544   1777474   206179             1                  0   
3420548    566068   206180             1                  0   
3420567    351920   206181             1                  0   
3420582    485448   206182             1                  0   
3420593     31169   206183             1                  0   
3420608    180109   206184             1                  6   
3420613   2168901   206185             1                  5   
3420624   1831589   206186             1                  6   
3420628    474057   206187             1                  2   
3420663   1944921   206188             1                  4   
3420671   2966736   206189             1                  4   
3420678   1955107   206190             1                  6   
3420687    936117   206191             1                  6   
3420693   3242108   206192             1                  6   
3420708    313609   206193             1                  5   
3420750   3395125   206194             1                  1   
3420762    926652   206195             1                  3   
3420782    825502   206196             1                  6   
3420787   3325493   206197             1                  2   
3420811   2154947   206198             1                  0   
3420819   3302466   206199             1                  2   
3420839   2508134   206200             1                  5   
3420863     28210   206201             1                  6   
3420896    799173   206202             1                  2   
3420919   1991850   206203             1                  1   
3420925   1438269   206204             1                  1   
3420930    969311   206205             1                  4   
3420934   3189322   206206             1                  3   
3421002   2166133   206207             1                  6   
3421019   2227043   206208             1                  1   
3421069   3154581   206209             1                  3   

         order_hour_of_day  days_since_last_order  
3420283                  7                    NaN  
3420294                 11                    NaN  
3420310                 10                    NaN  
3420329                  7                    NaN  
3420335                 12                    NaN  
3420341                 18                    NaN  
3420369                 16                    NaN  
3420389                  7                    NaN  
3420394                 15                    NaN  
3420407                 13                    NaN  
3420416                 15                    NaN  
3420423                 21                    NaN  
3420430                 17                    NaN  
3420437                 16                    NaN  
3420441        

Observation: based on head and tail output, all of the NaN missing values for days_since_last_order have a order number of 1, meaning that they are first time customers.

The max value in the 'user_id' column is 206209, which matches the number of missing values since these are each a new customer

## 6 Address changes for missing values

Because all of the missing values correspond to a new customer, these are not inaccurate observations.

Two options are to leave the 206209 values as NaN or to replace all NaN with 0. 

While leaving the values as NaN would still allow for descriptive statistics in python, but could lead to issues further down with creating visualizations.

Replacing NaN with 0 would help with visualizations, and it would skew descriptive stats. Howeve it could also be used to show trends in new vs. established customer bases.

Both NaN and 0 values could be used as indications of new customers.

These missing values are important information and must be included in the data set.

At this point, I chose to create a new data frame in which I replace NaN with 0, while also keeping df_ords df using NaN until I have more information about what stakeholders need in the analysis. 

In [67]:
# Create new data frame where all NaN (new customer days_since_last_order) are listed as '0' rather than NaN

df_ords_new_cust_values = df_ords.fillna(0)

In [68]:
# check new data frame for replacement of null values

df_ords_new_cust_values.isnull().sum()

# accurate replacement

order_id                 0
user_id                  0
order_number             0
order_day_of_week        0
order_hour_of_day        0
days_since_last_order    0
dtype: int64

Note: based on next steps, may need to complete further checks on both df_ords_new_cust_values and df_ords prior to saving each data frame

## 7 Run check for duplicate values

As noted, if duplicates are found, will need to address them in df_ords_new_cust_values as well as df_ords
Can choose which df to utilize in analysis once stakeholders propose their needs and questions

In [69]:
# check for full duplicates in df_ords

df_dups = df_ords[df_ords.duplicated()]

In [70]:
# list of full duplicates

df_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

## 8 Address any duplicates found

Observation: no full duplicates found. Given meaning up data, any partial duplicates between columns are not worrisome since there are naturally duplicates across order_number, order_day_of_week, etc.

No need for futher action regarding duplicates

## 9 Export dataframes

Need to export df from products, orders, and also orders that replaced null values with 0

In [71]:
# Export cleaned and no duplicates products dataframe as new file

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'df_prods_checked.csv'))

In [72]:
# Export cleaned orders dataframe with NaN (unchanged) missing values for new customers in 'days_since_last_order'

df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'), index = False)

In [76]:
# Export cleaned orders dataframe with changed missing values from NaN to '0' for new customers in'days_since_last_order'

df_ords_new_cust_values.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'df_ords_checked_new_cust_values.csv'), index = False)